drivers/soundwire/bus.c

```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 	return ret;
 }
 
+static int sdw_slave_set_frequency(struct sdw_slave *slave)
+{
+	u32 mclk_freq = slave->bus->prop.mclk_freq;
+	u32 curr_freq = slave->bus->params.curr_dr_freq >> 1;
+	u8 base_offset = 0;
+	u8 base = 0;
+	u8 scale_index = 0;
+	u32 scale;
+	int ret;
+	int i;
+
+	/*
+	 * frequency base and scale registers are required for SDCA
+	 * devices. They may also be used for 1.2+/non-SDCA devices,
+	 * but we will need a DisCo property to cover this case
+	 */
+	if (!slave->id.class_id)
+		return 0;
+
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
+		base = SDW_SCP_BASE_CLOCK_19200000_HZ;
+		base_offset = 1;
+	}
+	if (!(24000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24000000_HZ;
+	if (!(24576000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24576000_HZ;
+	if (!(22579200 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_22579200_HZ;
+	if (!(32000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_32000000_HZ;
+
+	/* paranoia checks */
+	if (!base) {
+		dev_err(&slave->dev,
+			"Unsupported clock base\n");
+		return -EINVAL;
+	}
+
+	if (mclk_freq % curr_freq) {
+		dev_err(&slave->dev,
+			"MCLK and bus curr_freq are not multiple\n");
```

Comment:
Can we print frequency value to make it more instructive ?

Reply:
not being a native English speaker - the above sentence means the same as "MCLK is not a multiple of curr_freq?"

Reply:
fixed